source/common/stats/thread_local_store.cc

```diff
@@ -699,26 +760,83 @@ void ThreadLocalHistogramImpl::merge(histogram_t* target) {
   hist_clear(*other_histogram);
 }
 
-ParentHistogramImpl::ParentHistogramImpl(StatName name, Histogram::Unit unit, Store& parent,
-                                         TlsScope& tls_scope, StatName tag_extracted_name,
+ParentHistogramImpl::ParentHistogramImpl(StatName name, Histogram::Unit unit,
+                                         ThreadLocalStoreImpl& parent, StatName tag_extracted_name,
                                          const StatNameTagVector& stat_name_tags,
-                                         ConstSupportedBuckets& supported_buckets)
+                                         ConstSupportedBuckets& supported_buckets, uint64_t id)
     : MetricImpl(name, tag_extracted_name, stat_name_tags, parent.symbolTable()), unit_(unit),
-      parent_(parent), tls_scope_(tls_scope), interval_histogram_(hist_alloc()),
-      cumulative_histogram_(hist_alloc()),
+      parent_(parent), interval_histogram_(hist_alloc()), cumulative_histogram_(hist_alloc()),
       interval_statistics_(interval_histogram_, supported_buckets),
-      cumulative_statistics_(cumulative_histogram_, supported_buckets), merged_(false) {}
+      cumulative_statistics_(cumulative_histogram_, supported_buckets), merged_(false), id_(id) {}
 
 ParentHistogramImpl::~ParentHistogramImpl() {
+  parent_.releaseHistogramCrossThread(id_);
+  ASSERT(ref_count_ == 0);
   MetricImpl::clear(parent_.symbolTable());
   hist_free(interval_histogram_);
   hist_free(cumulative_histogram_);
 }
 
+void ParentHistogramImpl::incRefCount() { ++ref_count_; }
+
+bool ParentHistogramImpl::decRefCount() {
+  bool ret;
+  if (shutting_down_) {
+    ret = --ref_count_ == 0;
+  } else {
+    ret = parent_.decHistogramRefCount(*this, ref_count_);
+  }
+  return ret;
+}
+
+bool ThreadLocalStoreImpl::decHistogramRefCount(ParentHistogramImpl& hist,
+                                                std::atomic<uint32_t>& ref_count) {
+  // We must hold the store's histogram lock when decrementing the
+  // refcount. Otherwise another thread may simultaneously try to allocate the
+  // same name'd stat after we decrement it, and we'll wind up with a
+  // dtor/update race. To avoid this we must hold the lock until the stat is
+  // removed from the map.
+  Thread::LockGuard lock(hist_mutex_);
+  ASSERT(ref_count >= 1);
+  if (--ref_count == 0) {
+    if (!shutting_down_) {
+      const size_t count = histogram_set_.erase(hist.statName());
+      ASSERT(shutting_down_ || count == 1);
+    }
+    return true;
+  }
+  return false;
+}
+
+uint32_t ThreadLocalStoreImpl::numHistograms() const {
+  Thread::LockGuard lock(hist_mutex_);
+  return histogram_set_.size();
+}
+
+uint32_t ThreadLocalStoreImpl::numTlsHistogramsForTesting() const {
```

Comment:
I'm wondering why we need this? Can't we verify correctness by shutting down and making sure there are no leaks and/or destroying a scope, then making a new scope (no overlapping period), and verifying the data is fresh? If we need it can there be more comments as to why?

Reply:
I thought about this a bit. The problem is that if I remove the call to `thread_local_store_.releaseHistogramCrossThread(id_);` from `ParentHistogramImpl::~ParentHistogramImpl`, there will not be a functional issue I can easily test. That's because the key to the TLS histograms is the monotonically increasing histogram ID, so there will be no co-mingling of old and new TlsHistogram objects, even if they are named the same thing.

There will also not be a memory leak because the TlsCache owns its `histogram_cache_` and will itself be destructed. So it will just leak into a pool and be cleaned up when the threading is shut down.

I also don't think it's going to be that easy to do a memory bytes-allocated test, because at the level of tcmalloc, the various STL structures will retain their high-water capacity even after histograms are cleaned up. I could work around this in prod by swapping histogram_cache_ with a fresh entry when it empties out, but that seems like a hack.

Reply:
One thing I could do is friend the ThreadLocalStore class and move this code into the test...maybe that's easiest.